da/celestia/celestia.go

```diff
@@ -111,60 +127,101 @@ func (c *DataAvailabilityLayerClient) SubmitBlock(ctx context.Context, block *ty
+				DAHeight: dataLayerHeight,
+			},
+			DataAvailable: false,
+		}
+	}
+	err = c.rpc.Share.SharesAvailable(ctx, header.DAH)
+	if err != nil {
+		if err.Error() == ErrNotAvailable.Error() {
+			return da.ResultCheckBlock{
+				BaseResult: da.BaseResult{
+					Code:     da.StatusSuccess,
+					DAHeight: dataLayerHeight,
+				},
+				DataAvailable: false,
+			}
+		}
 		return da.ResultCheckBlock{
 			BaseResult: da.BaseResult{
 				Code:    da.StatusError,
 				Message: err.Error(),
 			},
 		}
 	}
-
 	return da.ResultCheckBlock{
 		BaseResult: da.BaseResult{
 			Code:     da.StatusSuccess,
 			DAHeight: dataLayerHeight,
 		},
-		DataAvailable: len(shares) > 0,
+		DataAvailable: true,
 	}
 }
 
 // RetrieveBlocks gets a batch of blocks from DA layer.
-func (c *DataAvailabilityLayerClient) RetrieveBlocks(
-	ctx context.Context, dataLayerHeight uint64,
-) da.ResultRetrieveBlocks {
-	data, err := c.client.NamespacedData(ctx, c.namespaceID, dataLayerHeight)
-	status := dataRequestErrorToStatus(err)
-	if status != da.StatusSuccess {
+func (c *DataAvailabilityLayerClient) RetrieveBlocks(ctx context.Context, dataLayerHeight uint64) da.ResultRetrieveBlocks {
+	header, err := c.rpc.Header.GetByHeight(ctx, dataLayerHeight)
+	if err != nil {
 		return da.ResultRetrieveBlocks{
 			BaseResult: da.BaseResult{
-				Code:    status,
+				Code:    da.StatusError,
 				Message: err.Error(),
 			},
 		}
 	}
-
-	blocks := make([]*types.Block, len(data))
-	for i, msg := range data {
-		var block pb.Block
-		err = proto.Unmarshal(msg, &block)
-		if err != nil {
-			c.logger.Error("failed to unmarshal block", "daHeight", dataLayerHeight, "position", i, "error", err)
-			continue
+	rows, err := c.rpc.Share.GetSharesByNamespace(ctx, header.DAH, c.namespaceID[:])
+	if err != nil {
+		var code da.StatusCode
+		if strings.Contains(err.Error(), da.ErrDataNotFound.Error()) || strings.Contains(err.Error(), da.ErrEDSNotFound.Error()) {
+			code = da.StatusNotFound
+		} else {
+			code = da.StatusError
 		}
-		blocks[i] = new(types.Block)
-		err := blocks[i].FromProto(&block)
-		if err != nil {
-			return da.ResultRetrieveBlocks{
-				BaseResult: da.BaseResult{
-					Code:    da.StatusError,
-					Message: err.Error(),
-				},
+		return da.ResultRetrieveBlocks{
+			BaseResult: da.BaseResult{
+				Code:    code,
+				Message: err.Error(),
+			},
+		}
+	}
+
+	size := 0
+	for _, row := range rows {
+		size += len(row.Shares)
+	}
+	blocks := make([]*types.Block, size)
+	for _, row := range rows {
+		for i, share := range row.Shares {
+			var block pb.Block
+			err = proto.Unmarshal(share, &block)
+			if err != nil {
+				c.logger.Error("failed to unmarshal block", "daHeight", dataLayerHeight, "position", i, "error", err)
+				continue
+			}
+			blocks[i] = new(types.Block)
+			err := blocks[i].FromProto(&block)
+			if err != nil {
+				return da.ResultRetrieveBlocks{
+					BaseResult: da.BaseResult{
+						Code:    da.StatusError,
+						Message: err.Error(),
+					},
+				}
```

Comment:
what's the rational here that we continue on Unmarshal errors but we return on proto errors?

Reply:
Not sure as this change is not made by me. Possibly because if it's unmarshalled it's a valid block, but with invalid header / signature? cc @gupadhyaya @nashqueue

Reply:
my guess is that, the sequencer could submit random bytes to DA and hence we can ignore it. if we get the required bytes that decode to Block, then we want them to be properly built. @tzdybal could add better context.

Reply:
So when unmarshaling fails, the blob is random bytes that we cannot parse. We skip random bytes. But when we can unmarshal it but the type inside the header is wrong, for example we want to throw that error. I think that was the reasoning behind it, but maybe @liamsi has an idea. 

It might be a good idea to continue on proto error as well and just log, because someone could post a good marshal but wrong proto and the chain would not be processed even though we could just skip those blocks.

Reply:
I think we need follow-up issue for this. It's not clear what's the best approach here. As we shouldn't just return an error in case of invalid block - this way it's an easy attack vector - you can put something wrong in namespace to discard all valid blocks at given height.

Reply:
follow up #1043

Reply:
Closing as to follow up.